### PR TITLE
Merge 4.05 into trunk

### PR DESCRIPTION
**Pull request by damiendoligez**
Note: you should only look at the last two commits, all the others are from the 4.05 branch.

Here are my notes of things to check for this merge:

@shindere
 * config/Makefile.mingw
 * config/Makefile.mingw64
   check -fno-tree-vrp (in the right variable?)

@shindere and @dra27
 * testsuite/tests/lib-unix/common/Makefile
   two patches landing on the same lines
   @shindere: check vs trunk  + check line 26: missing @ ?
   @dra27: check vs 4.05

@xavierleroy
 * otherlibs/unix/unixLabels.mli
   missing function map_file ? (todo in another PR)

@dra27
 * appveyor.yml
   check "branches" section

@xavierleroy
 * otherlibs/bigarray/*
 * otherlibs/unix
 * stdlib
 * testsuite
   I removed all the bigarray-related changes found in the 4.05 branch
   and kept the trunk's current state because the  branch had reverted
   the bigarray changes. Is this correct?

### CODE REVIEW

**Comment by xavierleroy** [comment]
> I removed all the bigarray-related changes found in the 4.05 branch
> and kept the trunk's current state because the branch had reverted
> the bigarray changes. Is this correct?

Yes it is.  The first batch of changes (from GPR #997) wasn't quite perfect so (in March 2017) it was reverted from 4.05 and further improved on trunk (via GPR #1077).

File: otherlibs/unix/unix.mli [hunk 1]
```diff
@@ -541,7 +541,7 @@ val map_file :
 
   [Invalid_argument] or [Failure] may be raised in cases where argument
   validation fails.
-  @since 4.05.0 *)
+  @since 4.06.0 *)
```
Comment (xavierleroy) [inline otherlibs/unix/unix.mli:544]: Good catch, thanks!

**Review by xavierleroy — APPROVED** [review]
That's a short diff for 172 commits, I'm relieved.  Looks good to me.

**Comment by xavierleroy** [comment]
Concerning `UnixLabels.map_file`: see GPR #1245 and thanks for the reminder.

**Comment by shindere** [comment]
Damien Doligez (2017/07/13 18:29 +0000):
> Here are my notes of things to check for this merge:
> 
> @shindere
>  * config/Makefile.mingw
>  * config/Makefile.mingw64
>    check -fno-tree-vrp (in the right variable?)

There is a deeper issue here. A few build variables had been removed
after the 4.05 branch was created, so that #1228 can't just be applied
as-is to trunk.

More precisely, in configure, at lines 881 and below,

case "$bytecc,$target" in

should be replaced by

case "$cc,$target" in

Line 884:

should be changed, too:

- If the option should be used only when building C files in the OCaml
source tree, then it should be added to the internal_cflags environment
variable.

- If the option should be used internally *and* passed to the C compiler
when invoked by ocamlc and ocamlopt, then it should be added to the
common_cflags environment variable.

The answer to your question regarding config/Makefile.mingw and
config/Makefile.mingw64 is in the same vein:

- In the first case (internal use only) the modification is okay as-is.

- In the second case, then the build variables OCAMLC_CFLAGS and
OCAMLOPT_CFLAGS should also be modified.

> @shindere and @dra27
>  * testsuite/tests/lib-unix/common/Makefile
>    two patches landing on the same lines
>    @shindere: check vs trunk  + check line 26: missing @ ?

Yes. IIRC I removed it for debugging and then commited it
unintentionnally and @dra27 restored it.

**Comment by damiendoligez** [comment]
@shindere I pushed a new commit to take your remarks into account. The compiler option is indeed internal, we want to use only for compiling the runtimes.

**Comment by shindere** [comment]
Damien Doligez (2017/07/17 12:02 -0700):
> @shindere I pushed a new commit to take your remarks into account.

Thanks!

> The compiler option is indeed internal, we want to use only for
> compiling the runtimes.

I'm sorry, I don't understand what's the difference between the runtimes
and third-party C sources for this option. Why is it sure that compiling
third-party C source files without this option won't result in this
SkyLake bug, whereas this option seems so important when it comes to
compiling the runtimes?

Silly question I guess, but I can't figure out by myself.

**Comment by gasche** [comment]
That's actually a good question. It is not easy a priori to know whether some code may or may not trigger the bug; we know that it is unlikely for most C codebases (otherwhise it would have been found earlier), and we know that it is triggered by the OCaml GC code (it seems that the gc-bit-marking manipulations done by the GC are more likely to be compiled by GC to the half-register operations that are necessary to trigger the issue, along with a very short loop).
This suggests that user C code that does use OCaml memory-related macros might also be at risk, but maybe most of the "dangerous" macros (those that get compiled into half-register operations) are internal to the runtime and not part of the user-exposed FFI?

**Comment by xavierleroy** [comment]
Whether to have -fno-tree-vrp as a "private" compiler option was already discussed and decided in #1228.  Let's try to make progress instead of rehashing this point.

**Comment by damiendoligez** [comment]
Rebased and merged without waiting for further review because git makes it painful to rebase merges.
Please review anyway, any problems will be fixed in a new (non-merge) PR.

**Pull request merged**